chain/pool/txpool.go

```diff
@@ -678,3 +707,109 @@ func (tp *TxnPool) GetNonceByTxnPool(addr common.Uint160) (uint64, error) {
 func shortHashToKey(shortHash []byte) string {
 	return string(shortHash)
 }
+
+// sync txn
+// update twtFingerprint
+func (tp *TxnPool) updateTwtFingerprint(txnHash common.Uint256) {
+	byteTxnHash := txnHash.ToArray()
+	if len(tp.twtFingerprint) == 0 {
+		tp.twtFingerprint = byteTxnHash
+	} else {
+		for i := 0; i < len(byteTxnHash); i++ {
+			tp.twtFingerprint[i] ^= byteTxnHash[i]
+		}
+	}
+}
+
+// when append new txn to pool, we add this txn pointer to txtWithTime map too.
+func (tp *TxnPool) AppendTwt(txn *transaction.Transaction) {
+
+	now := time.Now().UnixMilli()
+
+	twt := TxnWithTime{
+		ArriveTime: now,
+		Txn:        txn,
+	}
+
+	tp.twtMapMu.Lock()
+	tp.twtMap.Set(txn.Hash(), twt)
+	tp.updateTwtFingerprint(txn.Hash())
+	tp.twtMapMu.Unlock()
+
+	atomic.AddInt32(&tp.twtCount, 1)
+
+	return
+
+}
+
+// romve time over items from txn with time list
+func (tp *TxnPool) RemoveExpiredTwt() {
+
+	now := time.Now().UnixMilli()
+	expiredThreshold := now - MaxSyncTxnInterval
+
+	delKeys := make([]interface{}, 0)
+
+	pair := tp.twtMap.Oldest()
+	for pair != nil {
+		twt := pair.Value.(TxnWithTime)
+		if twt.ArriveTime < expiredThreshold { // expired
+			delKeys = append(delKeys, pair.Key)
+		} else {
+			break
+		}
+		pair = pair.Next()
+	}
+
+	tp.twtMapMu.Lock()
+	defer tp.twtMapMu.Unlock()
+
+	for _, key := range delKeys {
+		tp.twtMap.Delete(key)
+		tp.updateTwtFingerprint(key.(common.Uint256))
+	}
+
+}
+
+// remove submitted txn from txn with time list
+func (tp *TxnPool) removeFromTwtMap(txnsRemoved []*transaction.Transaction) {
+	var totalRemoved int32 = 0
+
+	tp.twtMapMu.Lock()
+	for _, txn := range txnsRemoved {
+		txnHash := txn.Hash()
+		_, ok := tp.twtMap.Load(txnHash)
+		if ok {
+			tp.twtMap.Delete(txnHash)
+			tp.updateTwtFingerprint(txnHash)
+			totalRemoved++
+		}
+	}
+	tp.twtMapMu.Unlock()
+
+	if totalRemoved > 0 {
+		atomic.AddInt32(&tp.twtCount, -totalRemoved)
```

Comment:
Don't use atomic here, protect twtCount with the lock

Reply:
yes. fixed.